src/test/java/org/elasticsearch/script/mustache/MustacheScriptEngineTest.java

```diff
@@ -98,14 +103,60 @@ public void testEscapeJson() throws IOException {
                 escaped.append(c);
             }
             StringWriter target = new StringWriter();
-            assertThat(escaped.toString(), equalTo(JsonEscapingMustacheFactory.escape(writer.toString(), target).toString()));
+            assertThat(escaped.toString(), equalTo(InternalMustacheFactory.escape(writer.toString(), target).toString()));
         }
     }
 
+    @Test @Repeat(iterations = 5)
+    public void testArrayAccess() throws Exception {
+        Map<String, Object> vars = new HashMap<>();
+        Object data = randomFrom(
+                new String[] { "foo", "bar" },
+                ImmutableList.of("foo", "bar"),
+                ImmutableSet.of("foo", "bar"));
+        vars.put("data", data);
+        Object output = engine.execute(engine.compile("{{data.0}} {{data.1}}"), vars);
+        assertThat(output, notNullValue());
+        assertThat(output, instanceOf(BytesReference.class));
+        BytesReference bytes = (BytesReference) output;
+        assertThat(bytes.toUtf8(), equalTo("foo bar"));
+    }
+
+    @Test @Repeat(iterations = 5)
+    public void testArrayInArrayAccess() throws Exception {
+        Map<String, Object> vars = new HashMap<>();
+        Object data = randomFrom(
+                new String[][] { new String[] {"foo", "bar" }},
+                ImmutableList.of(new String[] {"foo", "bar" }),
+                ImmutableSet.of(new String[] {"foo", "bar" })
+        );
+        vars.put("data", data);
+        Object output = engine.execute(engine.compile("{{data.0.0}} {{data.0.1}}"), vars);
+        assertThat(output, notNullValue());
+        assertThat(output, instanceOf(BytesReference.class));
+        BytesReference bytes = (BytesReference) output;
+        assertThat(bytes.toUtf8(), equalTo("foo bar"));
+    }
+
+    @Test @Repeat(iterations = 5)
+    public void testMapInArrayAccess() throws Exception {
+        Map<String, Object> vars = new HashMap<>();
+        Object data = randomFrom(
+                new Map[] { ImmutableMap.<String, Object>of("key", "foo"), ImmutableMap.<String, Object>of("key", "bar") },
+                ImmutableList.of(ImmutableMap.<String, Object>of("key", "foo"), ImmutableMap.<String, Object>of("key", "bar")),
+                ImmutableSet.of(ImmutableMap.<String, Object>of("key", "foo"), ImmutableMap.<String, Object>of("key", "bar")));
+        vars.put("data", data);
+        Object output = engine.execute(engine.compile("{{data.0.key}} {{data.1.key}}"), vars);
+        assertThat(output, notNullValue());
+        assertThat(output, instanceOf(BytesReference.class));
+        BytesReference bytes = (BytesReference) output;
+        assertThat(bytes.toUtf8(), equalTo("foo bar"));
+    }
+
```

Comment:
> such that the deafult mustache engine can be used in other places (plugins) with this 
> extended functionality.

Just out of curiosity: Do you have a specific plugin in mind that already has this need? Not sure how we go about documentation for plugin authors - if mustache engine re-use is one of the major goals here IMHO there should be some doc/example for it? 

I guess I've seen too much unused code lingering in projects in the past that turned out to be "not quite what we need" which is why I try so hard to find real use cases for verification.